server/src-lib/Hasura/RQL/DDL/Schema/Table.hs

```diff
@@ -180,11 +183,26 @@ instance FromJSON SetTableCustomFields where
 runSetTableCustomFieldsQV2 :: (CacheBuildM m, UserInfoM m) => SetTableCustomFields -> m EncJSON
 runSetTableCustomFieldsQV2 (SetTableCustomFields tableName rootFields columnNames) = do
   adminOnly
-  void $ askTabInfo tableName
+  fieldInfoMap <- _tiFieldInfoMap <$> askTabInfo tableName
+  validateWithExistingRelationships fieldInfoMap
   let tableConfig = TableConfig rootFields columnNames
   updateTableConfig tableName tableConfig
   buildSchemaCacheFor (MOTable tableName)
   return successMsg
+  where
+    validateWithExistingRelationships fields = do
```

Comment:
I don’t think I totally understand all the subtleties of this issue. Could you explain why `validateWithExistingRelationships` is different from/not a part of `validateTableConfig`?

Reply:
Sure. `validateTableConfig` happens when building `TableInfo`. At that time no relationships were added. `set_table_custom_fields` is applied for only tracked tables. So we should validate with existing relationships.

Reply:
Well, I also stuck with the thought while writing the above reply. What about validating the new relationships being added with existing custom column names? We should do that definitely. Will push a commit with that change soon.

Reply:
Putting my all thoughts below:

I have a simple schema

```plpgsql
CREATE TABLE "user"(
  id SERIAL PRIMARY KEY,
  name TEXT NOT NULL,
  age INTEGER
)
```
I tracked the above table with the following custom column names configuration.
```json
{
    "id": "userId",
    "name": "userName"
}
```
What could be the best way to validate a relationship with name `X` that is being defined? Currently, we only check with existing column names and relationships. We should definitely validate with custom column names (`userId` and `userName`). 
Also, I'm thinking about new class fields coming soon (computed fields and remote joins). I think we should improve the logic of function `checkForFieldConflict` and should change `validateWithExistingRelationships` to something like `validateWithFieldsOtherThanColumns`. 

I'll answer all these in the next commit.

Reply:
> Sure. `validateTableConfig` happens when initialising `TableInfo` of a table. At that time no relationships were added. Hence, we cannot do validation with relationships.

Got it, that makes sense! As a followup question: do you think we could consolidate this logic somehow? I know that each code path has slightly different inputs, and you have a firmer grasp on the details than I do, but it would be nice if we could share the logic between `validateTableConfig`, `validateWithNonColumnFields`, and `checkForFieldConflict` somehow so that it can’t get out of sync.

Reply:
Prior to [this](https://github.com/hasura/graphql-engine/commit/ed26da59a6d00b34e67707b754a9622848064e75) commit the server used to build the `TableInfo` on `track_table` query and add it to the cache later. Now `buildSchemaCacheFor` is used to construct `TableInfo` from scratch and add update it with other table metadata objects. Here we don't have luxury of having `TableConfig` before we call `buildSchemaCacheFor` to validate the input custom fields configuration. This forces the validation of custom fields configuration with existing column names to occur only in `buildSchemaCacheFor` code path.

In the `set_table_custom_fields` query API, `buildSchemaCacheFor` is being used to rebuild the `TableConfig`. Here, the validation for custom fields occur in two steps
1. With existing un-updated `TableInfo`, validate the custom column fields with existing non-columns (relationships, computed fields as of now).
2. We can validate the input custom column fields with existing columns here. But, it doesn't make sense since the validation of custom column fields occurs in `buildSchemaCacheFor` code path.

After a bit of thinking,
1. Scrap `validateTableConfig` function and break it into two isolated logics; `validateCustomRootFields` and `validateWithColumns`.
2. Use `validateCustomRootFields` directly in `trackExistingTableOrViewP2` and `runSetTableCustomFields`
3. Use `validateWithColumns` only in `buildTableCache`